File: app/src/main/java/com/prey/services/AppAccessibilityService.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2020 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.services;

import android.accessibilityservice.AccessibilityService;
import android.content.Intent;
import android.os.Build;
import android.view.accessibility.AccessibilityEvent;

import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.activities.PasswordHtmlActivity;
import com.prey.activities.PasswordNativeActivity;

public class AppAccessibilityService extends AccessibilityService {

    @Override
    public void onCreate() {
        super.onCreate();
        PreyLogger.d("AppAccessibilityService onCreate");
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent accessibilityEvent) {
        try {
            String unlockPass=PreyConfig.getPreyConfig(getApplicationContext()).getUnlockPass();
            boolean isLock = unlockPass!=null && !"".equals(unlockPass);
            if ( isLock ) {
                PreyLogger.d("acc 1");
                if (accessibilityEvent != null && accessibilityEvent.getPackageName() != null) {
                    String charSequence = accessibilityEvent.getPackageName() != null ? accessibilityEvent.getPackageName().toString() : null;
                    if ("com.prey".equals(charSequence) || "android".equals(charSequence)) {
                    } else {
                        if(isLock) {
                            PreyLogger.d("acc 2");
                            Intent intentPasswordActivity = null;
                            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                                intentPasswordActivity = new Intent(getApplicationContext(), PasswordHtmlActivity.class);
                            }else{
                                intentPasswordActivity = new Intent(getApplicationContext(), PasswordNativeActivity.class);
                            }
                            intentPasswordActivity.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            getApplicationContext().startActivity(intentPasswordActivity);
                        }
                    }
                }
            }
        }catch (Exception e){
            PreyLogger.e("Error onAccessibilityEvent:"+e.getMessage(),e);
        }
    }

    @Override
    public void onInterrupt() {
    }

}

File: app/src/main/java/com/prey/services/AppAccessibilityService.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.services

import android.accessibilityservice.AccessibilityService
import android.content.Intent
import android.os.Build
import android.view.accessibility.AccessibilityEvent

import com.prey.activities.PasswordHtmlActivity
import com.prey.activities.PasswordNativeActivity
import com.prey.PreyConfig
import com.prey.PreyLogger

/**
 * AppAccessibilityService is an AccessibilityService that handles accessibility events.
 * It checks if the device is locked and if so, it starts the password activity.
 */
class AppAccessibilityService : AccessibilityService() {

    /**
     * Called when the service is created.
     */
    override fun onCreate() {
        super.onCreate()
        PreyLogger.d("AppAccessibilityService onCreate")
    }

    /**
     * Called when an accessibility event occurs.
     *
     * @param accessibilityEvent The accessibility event that occurred.
     */
    override fun onAccessibilityEvent(accessibilityEvent: AccessibilityEvent) {
        try {
            val unlockPass = PreyConfig.getInstance(applicationContext).getUnlockPass()
            val isLock = unlockPass != null && "" != unlockPass
            if (isLock) {
                PreyLogger.d("acc 1")
                if (accessibilityEvent?.packageName != null) {
                    val charSequence =
                        if (accessibilityEvent.packageName != null) accessibilityEvent.packageName.toString() else null
                    if ("com.prey" == charSequence || "android" == charSequence) {
                    } else {
                        if (isLock) {
                            PreyLogger.d("acc 2")
                            var intentPasswordActivity: Intent? = null
                            intentPasswordActivity =
                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                                    Intent(applicationContext, PasswordHtmlActivity::class.java)
                                } else {
                                    Intent(
                                        applicationContext,
                                        PasswordNativeActivity::class.java
                                    )
                                }
                            intentPasswordActivity.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                            applicationContext.startActivity(intentPasswordActivity)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            PreyLogger.e("Error onAccessibilityEvent:${e.message}" , e)
        }
    }

    /**
     * Called when the service is interrupted.
     */
    override fun onInterrupt() {
    }
}